src/IR/TypeInference.hs

```diff
@@ -4,72 +4,203 @@ For now, this module implements a very crude form of pure, syntax-directed type
-inferExpr (I.Lit i@(I.LitIntegral _) t      ) = I.Lit i $ int 32 <> t
-inferExpr (I.App a b t                      ) = I.App a' b' (retTy <> t)
- where
-  (a', b') = (inferExpr a, inferExpr b)
-  retTy    = snd $ fromJust $ dearrow $ extract a'
-inferExpr e = e
+inferProgram p = runInferFn $ do
+  defs' <- inferTop $ I.programDefs p
+  return $ I.Program { I.programDefs  = defs'
+                     , I.programEntry = I.programEntry p
+                     , I.typeDefs     = [] -- TODO: something with I.typeDefs p
+                     }
+
+inferTop :: [(I.VarId, I.Expr Ann.Type)] -> InferFn [(I.VarId, I.Expr Classes.Type)]
+inferTop [] = return []
+inferTop ((v, e):xs) = do
+  e' <- inferExpr e
+  insertVar v (extract e')
+  xs' <- inferTop xs
+  return $ (v, e'):xs'
+
+inferExpr :: I.Expr Ann.Type -> InferFn (I.Expr Classes.Type)
+inferExpr (I.Var v t) = do
+  record <- lookupVar v
+  case record of
+    Nothing -> do
+      t' <- anns2Class t
+      insertVar v t'
+      return $ I.Var v t'
+    Just t' -> return $ I.Var v t'
+inferExpr e@(I.Lambda v b t) = do
+  b' <- withNewScope $ withVty e v t >> inferExpr b
+  t' <- anns2Class t
+  return $ I.Lambda v b' t'
+inferExpr (I.Let vs b _) = do
+  (vs', b') <- withNewScope $ localInferExpr vs b
+  return $ I.Let vs' b' (extract b')
+inferExpr e@I.Prim {} = inferPrim e
+inferExpr (I.Lit I.LitEvent _) = return $ I.Lit I.LitEvent unit
+inferExpr (I.Lit i@(I.LitIntegral _) _) = return $ I.Lit i (int 32)
+inferExpr e@(I.App a b _) = do
+  a' <- inferExpr a
+  b' <- inferExpr b
+  case dearrow $ extract a' of
+    Just (t1, t2) ->
+      if t1 == extract a'
+        then throwError $ Compiler.TypeError $ "App expression has inconsistent type: " ++ show e
+        else return $ I.App a' b' t2
+    Nothing -> throwError $ Compiler.TypeError $ "Unable to type App expression: " ++ show e
+inferExpr e = throwError $ Compiler.TypeError $ "Unable to type unknown expression: " ++ show e
+
+inferPrim :: I.Expr Ann.Type -> InferFn (I.Expr Classes.Type)
+inferPrim (I.Prim I.New [e] _) = do
+  e' <- inferExpr e
+  return $ I.Prim I.New [e'] $ ref $ extract e'
+inferPrim (I.Prim I.Deref [e] _) = do
+  e' <- inferExpr e
+  return $ I.Prim I.Deref [e'] $ fromJust $ deref $ extract e'
+inferPrim (I.Prim I.Dup [e] _) = do
+  e' <- inferExpr e
+  return $ I.Prim I.Dup [e'] $ extract e'
+inferPrim e@(I.Prim I.Assign [lhs, rhs] _) = do
+  lhs' <- inferExpr lhs
+  rhs' <- inferExpr rhs
+  let rrty = ref $ extract rhs'
+  if extract lhs' == rrty
+    then do
+      case lhs' of
+        I.Var v _ -> insertVar v rrty
+        _ -> return ()
+      return $ I.Prim I.Assign [rrty <$ lhs', rhs'] unit
+    else throwError $ Compiler.TypeError $ "Assign expression has inconsistent type: " ++ show e
+inferPrim e@(I.Prim I.After [del, lhs, rhs] _) = do
+  del' <- inferExpr del
+  rhs' <- inferExpr rhs
+  lhs' <- inferExpr lhs
+  let rrty = ref $ extract rhs'
+  if extract del' == int 32 && extract lhs' == rrty
```

Comment:
Delay should probably actually be 64-bit int. Can you change that hard-coded int size everywhere?